server/routes/TODOItems.js

```diff
@@ -0,0 +1,153 @@
+const express = require('express');
+const { TODOItem, validate } = require("../models/TODOItem.model");
+const Category = require('../models/Category.model');
+const router = express.Router();
+
+// retrieves the list of TODOItems associated with a particular UserID.
+// Filtering by categoryID can be applied to the output list.
+router.get('/fetchAllItems/:userID/:categoryID?', async (req, res) => {
+  try {
+    // Get list of TODOItems that are associated with current user
+    const filter = {
+      userID: req.params.userID
+    };
+
+    // add to filter object only if user provided additional filter attributes
+    if (req.params.categoryID) {
+      filter.category = req.params.categoryID;
+    }
+
+    let outputList = await TODOItem
+      .find(filter)
+      .select({ _id: 1, title: 1, startDate: 1, endDate: 1 });
+
+    // sort the todoItems by their startDate attribute
+    outputList = outputList.sort((todoOne, todoTwo) => {
+      const dateOne = new Date(todoOne.startDate);
+      const dateTwo = new Date(todoTwo.startDate);
+      return dateOne - dateTwo;
+    });
+
+    res.send(outputList);
+  } catch (error) {
+    console.error('Error fetching todo items:', error);
+    res.status(500).send('An error occurred while fetching todo items: ' + error.message);
+  }
+});
+
+router.get('/:itemID', async (req, res) => {
+  try {
+    const item = await TODOItem
+      .findById(req.params.itemID)
+      .populate('category')
+      .select("-userID");
+
+    if (!item) {
+      return res.status(404).send('The item with the given ID was not found.');
+    }
+
+    res.send(item);
+  } catch (error) {
+    console.error('Error fetching todo item:', error);
+    res.status(500).send('An error occurred while fetching the todo item: ' + error.message);
+  }
+});
+
+router.post('/', async (req, res) => {
+  try {
+    const { error } = validate(req.body);
+
+    if (error) {
+      return res.status(400).send(error.details[0].message);
+    }
+
+    let category = await Category
+      .findOne({ category: req.body.category });
+
+    // Create new Category document if user introduced a new category
```

Comment:
I find it odd that the only way to create a new Category is to create a TODO item. If a user just wanted to add multiple categories for future use the workflow would be needlessly lengthy,

Reply:
I think creating a Category while adding/editing TODOs makes more sense. Simply creating Categories without an associated TODO seems to be less often use case